Add  beforeUpdateQuantity to BeforeLineItemQuantityChangedEvent to allow plugins to understand if quantity increases or decreases

luminalpark opened the pull request:
### 1. Why is this change necessary?
`BeforeLineItemQuantityChangedEvent `should expose a `beforeUpdateQuantity` property , to allow plugins to understand if quantity increases or decreases. 

### 2. What does this change do, exactly?
Adds a new `beforeUpdateQuantity` property to `BeforeLineItemQuantityChangedEvent` to allow plugins to understand if quantity is being increased or decreased

### 3. Describe each step to reproduce the issue or behaviour.


### 4. Please link to the relevant issues (if any).


### 5. Checklist

- [x] I have rebased my changes to remove merge conflicts
- [ ] I have written tests and verified that they fail without my change
- [x] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-implement-new-changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x ] I have read the contribution requirements and fulfil them.

tobiasberge committed:
NEXT-28114 - Improve pagination accessibility

CR0YD committed:
NEXT-37109 - Added timezone hint to datepicker component

tajespasarela committed:
NEXT-37540 - Add eslint rule for explicit emits definition.

phizab committed:
NEXT-37684 - Fix updating thumbnails in strict mode
fixes #3836

philipreinken committed:
NEXT-37583 - Match CVE IDs in release description

nguyenytran committed:
NEXT-37571 - Fix VAT format pattern does not work correct

aragon999 committed:
NEXT-37699 - fix: Add visibility to block defaults
fixes #3842

M-arcus committed:
NEXT-37709 - Add max length constant of text to ElasticSearch indexing
fixes #3845

vienthuong committed:
NEXT-37659 - Fix set group scope discount calculation

tajespasarela committed:
NEXT-37540 - Fix code for the new eslint rule for explicit emits definition.

jozsefdamokos committed:
NEXT-37559 - Remove language dropdown from Settings > Basic Information

mstegmeyer committed:
NEXT-37600 - Fix endless loop in StateMachineRegistry

shyim committed:
NEXT-37673 - Don't save cart on any request when product is in cart

shyim committed:
NEXT-37673 - Dont save cart on any request when error is in cart

Phil23 committed:
NEXT-33689 - Improved registration form accessibility

luminalpark committed:
updateLineItem emits BeforeLineItemQuantityChangedEvent event before changing lineItem quantity

luminalpark committed:
updateLineItem emits BeforeLineItemQuantityChangedEvent event before changing lineItem quantity

luminalpark committed:
updateLineItem emits BeforeLineItemQuantityChangedEvent event before changing lineItem quantity

luminalpark committed:
updateLineItem emits BeforeLineItemQuantityChangedEvent event before changing lineItem quantity - rework

luminalpark committed:
created changelog file

luminalpark committed:
created changelog file

luminalpark committed:
updateLineItem emits BeforeLineItemQuantityChangedEvent event before changing lineItem quantity

luminalpark committed:
created changelog file

keulinho committed:
NEXT-31700 - Rely on serializer interface in message queue size listener

luminalpark committed:
Merge remote-tracking branch 'origin/BeforeLineItemQuantityChangedEvent' into BeforeLineItemQuantityChangedEvent

aragon999 commented:
While I understand the purpose of this change, this will break all plugins which rely on the fact, that `$lineItem->getQuantity()` will yield the changed quantity.

luminalpark commented:
> While I understand the purpose of this change, this will break all plugins which rely on the fact, that `$lineItem->getQuantity()` will yield the changed quantity.

Thanks a lot for sharing your view on my first PR! You're right.
The event dispatch could remain unmodified, I could add a `beforeUpdateQuantity` to the event to keep the desired functionality. What do you think?

luminalpark committed:
Merge remote-tracking branch 'origin/BeforeLineItemQuantityChangedEvent' into BeforeLineItemQuantityChangedEvent

luminalpark committed:
Passing initial lineItem quantity to BeforeLineItemQuantityChangedEvent event, so plugins can request it and understand if lineItem quantity has been increased or decreased.
Renamed and updated changelog file.

luminalpark committed:
Reverted LineItemFactoryRegistry to original, passing initial lineItem quantity to BeforeLineItemQuantityChangedEvent event, so plugins can request it and understand if lineItem quantity has been increased or decreased.
Renamed and updated changelog file.

luminalpark committed:
Passing initial lineItem quantity to BeforeLineItemQuantityChangedEvent event, so plugins can request it and understand if lineItem quantity has been increased or decreased.
Renamed and updated changelog file.

luminalpark commented:
I reverted part of the code, and added a beforeUpdateQuantity property to let plugins handle the delta: `$event->getBeforeUpdateQuantity() - $event->getLineItem()->getQuantity()`

keulinho commented on src/Core/Checkout/Cart/Event/BeforeLineItemQuantityChangedEvent.php:
```diff
@@ -28,10 +33,12 @@ class BeforeLineItemQuantityChangedEvent implements ShopwareSalesChannelEvent, C
     protected $salesChannelContext;
 
     public function __construct(
+        int $beforeUpdateQuantity,
```
this is a breaking change, maybe add a separate setter for that or directly use a public property?

keulinho commented on src/Core/Checkout/Cart/Event/BeforeLineItemQuantityChangedEvent.php:
```diff
@@ -12,6 +12,11 @@
+    /**
+     * @var int
+     */
+    protected $beforeUpdateQuantity;
```
Please use typed properties for all new code

phuoccx committed:
NEXT-37161 - Dispatch event after update stock and invalidate detail route

Commit:
NEXT-37567 - Update timestamp and signature for iframe sources

SilasHachmann committed:
NEXT-36792 - Validate array-type in ResponseFields

keulinho committed:
NEXT-37410 - Fix price validation in custom fields

Commit:
NEXT-37684 - Refactored ThumbnailServiceTest and provided test for new state

Commit:
NEXT-35606 - Disable compat groups S01

MarcelSchmaeing committed:
NEXT-00000 - remove old issue tracker occurences in the code

CR0YD committed:
NEXT-37412 - Added customer data to CustomerDeletedEvent

Bird87ZA commented:
@luminalpark Thanks for opening the PR.

Could you also write tests for your changes?

m-waes committed:
NEXT-37724 - Add composer name to `plugin:list` output for quicker access
fixes #4461

OliverSkroblin committed:
NEXT-31669 - Fix esi caching

philipreinken committed:
NEXT-37189 - Merge back v6.6.4.1

frobel committed:
NEXT-37715 - Make promotionId writable for Admin API

Commit:
NEXT-37699 - Added ticket number to changelog

shyim committed:
NEXT-37716 - Remove locust

philipreinken committed:
NEXT-37583 - Merge back v6.6.5.1

keulinho committed:
NEXT-37685 - Fix coverage per area job

philipreinken committed:
NEXT-00000 - Update releases.json

shyim committed:
NEXT-37724 - Update issue ticket number

philipreinken committed:
NEXT-37189 - Remove test directory that was added again by mistake

SilasHachmann committed:
NEXT-37579 - Exception for value

shyim committed:
NEXT-00000 - Ignore danger errors in PhpStan composer package

shyim committed:
NEXT-37733 - Improve property search

jleifeld committed:
NEXT-36650 - disable vue compat group 81

jleifeld committed:
NEXT-35606 - disable-multiple-vue-compat-groups-part-2

luminalpark committed:
Merge remote-tracking branch 'origin/BeforeLineItemQuantityChangedEvent' into BeforeLineItemQuantityChangedEvent

luminalpark committed:
Setting $beforeUpdateQuantity in BeforeLineItemQuantityChangedEvent, so plugins can request it and understand if lineItem quantity has been increased or decreased.
Updated changelog file.

luminalpark committed:
Setting $beforeUpdateQuantity in BeforeLineItemQuantityChangedEvent, so plugins can request it and understand if lineItem quantity has been increased or decreased.
Updated changelog file.

luminalpark committed:
Merge branch 'trunk' into BeforeLineItemQuantityChangedEvent

luminalpark committed:
Adding test

luminalpark committed:
Adding test

luminalpark committed:
Merge remote-tracking branch 'origin/BeforeLineItemQuantityChangedEvent' into BeforeLineItemQuantityChangedEvent

luminalpark committed:
Removed superfluous Phpdoc
Fixed linting issue

luminalpark committed:
Removed superfluous Phpdoc
Fixed linting issue

luminalpark committed:
Merge branch 'trunk' into BeforeLineItemQuantityChangedEvent

luminalpark committed:
Merge branch 'trunk' into BeforeLineItemQuantityChangedEvent

luminalpark committed:
Fixed missing issue number in changelog

luminalpark committed:
Fixed missing issue number in changelog

luminalpark committed:
Merge remote-tracking branch 'origin/BeforeLineItemQuantityChangedEvent' into BeforeLineItemQuantityChangedEvent

luminalpark committed:
Merge branch 'shopware:trunk' into BeforeLineItemQuantityChangedEvent
